File: graphs/course_schedule.cpp
// Approach : the problem can be mapped to simply finding whether the cycle exits in a directed graph
// Based on the prerequisites, we can make adjacency list and indegree array 
// indegree of a node indicates how many courses are prerequisites for this particular course , 
// indegree = 0 indicated no prereuisites and hence that course can be taken, 
// when a course is taken it opens up the way for other course to be taken, hence indegree of other course is decreased and 
// that too is pushed onto the the queue if no prerequisites are left, 
// this way just count the number of courses that can be taken(elem on the queue) and compare with the number given 

bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    vector<int>indegree(numCourses,0);
    vector<int>adj[numCourses];

    // making adjacency list and initializing indegree to have a clear view of the dependency 
    for(auto i : prerequisites){
        adj[i[1]].push_back(i[0]);
        indegree[i[0]]++;
    }

    queue<int>q ; // psuh all the courses with no prerequsites on the queue , indegree indicates prerequisites
    for(int i = 0 ; i < numCourses ; i++){
        if (indegree[i] == 0)  
            q.push(i);
    }
    int count = 0 ;
    while(!q.empty()){ // keep counting the elements on the queue and updating the adjacents' indegree
        int node = q.front();
        q.pop();
        count ++ ;

        for(auto x: adj[node]){ // update the prerequisites for others if this course is taken
            indegree[x]--;
            if (indegree[x] == 0) // if no prerequisites, then push on to the queue
                q.push(x);
        }
    }

    return count == numCourses ; // count if finally all the course taken equals the total number of courses or not 
}

// TC = O(V+E);
// SC = O(V);

File: graphs/go/course_schedule.go
// Idea is to make use of dfs here , we first of all make the adjacency list from
// the given set of requirements, once we have that we can simply make use of the
// fact that if we have a cycle in the dependency list for the courses, we simply
// wont be able to meet the requirements, so the problem comes down to just
// detecting the cycle in a directed graph, we make use of dfs as a first approach here
// TC = O(V+E) , same as dfs
// SC = O(V) , for extra space used , rec stackt too in wc

func canFinish(count int, pre [][]int) bool {
	// adj matrix banao pehle
	adj := make(map[int][]int, count)
	// [1,0] means you need to take 0 before 1

	for _, iter := range pre {
		course, req := iter[0], iter[1]
		adj[course] = append(adj[course], req)
	}

	vis := make([]int, count)

	for i := 0; i < count; i++ {
		if vis[i] == 0 && loopon(i, vis, adj) {
			return false
		}
	}
	return true
}

func loopon(node int, vis []int, adj map[int][]int) bool {
	vis[node] = 1

	for _, u := range adj[node] {
		if vis[u] == 0 && loopon(u, vis, adj) {
			return true // if loop is here
		} else if vis[u] == 1 {
			return true
		}
	}
	vis[node] = 2
	return false
} 